lsm.go

```diff
@@ -0,0 +1,352 @@
+package frostdb
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/apache/arrow/go/v14/arrow"
+	"github.com/apache/arrow/go/v14/arrow/util"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/polarsignals/frostdb/dynparquet"
+	"github.com/polarsignals/frostdb/parts"
+	"github.com/polarsignals/frostdb/query/logicalplan"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// LSM is a log-structured merge-tree like index. It is implemented as a single linked list of parts.
+//
+// Arrow records are always added to the L0 list. When a list reaches it's configured max size it is compacted
+// calling the levels Compact function and is then added as a new part to the next level.
+//
+// [L0]->[record]->[record]->[L1]->[record/parquet]->[record/parquet] etc.
+type LSM struct {
+	sync.RWMutex
+	compacting *atomic.Bool
+
+	prefix  string
+	levels  *Node
```

Comment:
What is the motivation behind structuring this index as a single linked list? Were there any other options you considered?

Reply:
I looked at a few other options but ultimately I wanted a lockless implementation and I wanted it to be dead simple since we did not need most of the features of an lsm. So start simple and build complexity as we need it.